test/puppetlabs/puppetdb/performance/cli.clj

```diff
@@ -0,0 +1,134 @@
+       (map (fn [v]
+              (let [m (meta v)]
+               (-> m
+                   (select-keys [:improve :regress :max-dev])
+                   (assoc :name (subs (str v) 2))))))
+       (into #{})))
+
+(defn run-test [var-name]
+  (require-tests)
+  (pr ((find-var (symbol var-name)))))
+
+(defn resolve-ref [ref]
+  (s/trim (:out (sh "git" "rev-parse" ref))))
+
+(defn git-export [ref work-tree]
+  (sh "sh" "-c" (str "git archive --format=tar --prefix=" work-tree "/ " ref " | tar xf -")))
+
+(defn run-all-tests-in [dir]
+  (let [results-file (str dir "/perf-result.edn")]
+    (if (fs/exists? results-file)
+      (do
+        (println "Reading cached results from" results-file)
+        (edn/read-string (slurp results-file)))
+      (do
+        (println "Running tests in" dir)
+        (let [tests (-> (sh "lein" "perf-test" "list" :dir dir) :out edn/read-string)
+              result-map (->> tests
+                              (map (fn [t]
+                                     (println " -" (:name t))
+                                     (let [out (:out (sh "lein" "perf-test" "run" (:name t)
+                                                         :dir dir))
+                                           result (edn/read-string {:default (fn [tag x] x)} out)]
+                                       (assoc t :result result))))
+                              (into #{}))]
+          (spit results-file (pr-str result-map))
+          result-map)))))
+
+(defn- outer-join [xrel yrel ks]
+  (let [keyed-x (set/index xrel ks)
+        keyed-y (set/index yrel ks)
+        all-keys (set/union (keys keyed-x) (keys keyed-y))]
+    (->> all-keys
+         (map (fn [k]
+                (let [xrow (first (get keyed-x k))
+                      yrow (first (get keyed-y k))]
+                  (merge xrow yrow))))
+         (into #{}))))
+
+(defn check-combined-test-result [{:keys [regress improve max-dev new old] :as test}]
+  ;; we write down ms, as is typical in clojure, but criterium gives us seconds
+  (let [regress (/ regress 1000)
+        improve (/ improve 1000)
+        max-dev (/ max-dev 1000)
+        old-time (first (:mean old))
+        new-time (first (:mean new))
+        dt (- new-time old-time)
+        variance (first (:variance new))]
+    (assoc test :errors
+           (remove nil?
+                   [(when (>= dt regress) (str "Performance regression of " dt
+                                               " is past threshold of " regress
+                                               " (" old-time "s -> " new-time "s)"))
+                    (when (<= (- improve) dt) (str "Suspicious performance improvement: "
+                                                   dt " is past threshold of " regress
+                                                   " (" old-time "s -> " new-time "s)"))
+                    (when (> variance max-dev) (str "Variance of " variance " is outside specified bound ("
```

Comment:
This is why I thought the `variance` field was actually standard deviation: https://github.com/hugoduncan/criterium/blob/develop/src/criterium/core.clj#L976

I haven't actually looked at how it's computed though.